### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized knn/perc model with integer arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#325]
Public recipe.train().optimizer(adamw).lr(0.0001).seed(47133611751318).epochs(1).log(all).stop(0.0).int(8).run() on backend:nv0 stalls at 65s before epoch arithmetic. Earliest failure is recipe.train().run() at compile_model -> NativeModelIr::from_graph -> emit -> compile_native_artifact on the NVIDIA ptx/ptxas branch (recipe.rs ~2852-2896). Model: .knn(3).norm(batch).qi(3).k.l .perc(8).prelu().iq(4).nl .loss(mae). The knn estimator lowers through fit_knn to Predictor (Primitive::Predictor) with quantized batch-normalized nodes, and .int(8) selects Compute::INT8, producing a native-artifact key not represented by 282 (svm/kmeans predictor nv0), 303 (KNN predictor CPU), or 321 (perc/cbst nv0). The numeric subfolder loader is ignored at the stall point; small data (~64 folders) with 3-NN and width-8 perc passes justify sub-second arithmetic, so 65s is compilation overhead, not throughput.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `19847f0f2e18e7dc`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"new","issue":0,"title":"NVIDIA nv0 training stalls beyond 60s on quantized knn/perc model with integer arithmetic","body":"Public recipe.train().optimizer(adamw).lr(0.0001).seed(47133611751318).epochs(1).log(all).stop(0.0).int(8).run() on backend:nv0 stalls at 65s before epoch arithmetic. Earliest failure is recipe.train().run() at compile_model -> NativeModelIr::from_graph -> emit -> compile_native_artifact on the NVIDIA ptx/ptxas branch (recipe.rs ~2852-2896). Model: .knn(3).norm(batch).qi(3).k.l .perc(8).prelu().iq(4).nl .loss(mae). The knn estimator lowers through fit_knn to Predictor (Primitive::Predictor) with quantized batch-normalized nodes, and .int(8) selects Compute::INT8, producing a native-artifact key not represented by 282 (svm/kmeans predictor nv0), 303 (KNN predictor CPU), or 321 (perc/cbst nv0). The numeric subfolder loader is ignored at the stall point; small data (~64 folders) with 3-NN and width-8 perc passes justify sub-second arithmetic, so 65s is compilation overhead, not throughput.","rationale":"Searched open nv0 performance issues (271, 282, 291, 293, 302, 308, 314, 315, 317, 319, 321) and CPU KNN issues (303, 254, 255). Read issue 282 (svm/kmeans predictor nv0 native compilation) and 303 (CPU KNN predictor native compilation) fully; both reach compile_native_artifact but represent different predictor primitives and backends. Inspected recipe.rs compile_model (2868) and compile_native_artifact (2828) showing the NVIDIA branch stalls at ptx/ptxas emission. The reproduction uses .knn(3) (fit_knn -> Predictor) plus .perc(8) block with .qi(3).k.l / .iq(4).nl under Compute::INT8 (.int(8)), a graph combination absent from existing issues. Because the earliest failing operation and implementation path (quantized Predictor lowering to native artifact compilation on nv0) are not covered by any open issue for this model path, this is a new defect, not a comment on 282 or 303."}
```

## Deterministic failure packet

kind=performance
id=19847f0f2e18e7dc
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10002 next:10003 composition:47133611751318
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 65 seconds
measurement=elapsed_seconds:65 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-10002-304.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/numeric/sample_subfolders")
		.target("target");
	let model = recipe.model()
		.knn(3).norm(batch).qi(3).k.l
		.perc(8).prelu().iq(4).nl
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(47133611751318)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.int(8)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #432.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #432: Verify current NVIDIA KNN execution</summary>

Closes #325

## Verification

- Ran the reported numeric subfolder data, KNN and perceptron model, quantization, INT8 arithmetic, optimizer, seed, and one-epoch training path through `./target/release/recipe --device benji:nv0`. Only the saved bundle path changed from `/tmp` to `/home/nate/.cache/recipe/issues/325/bundle.ogdl` because `/tmp` is tmpfs.
- The current loader produced 64 rows with an input width of 477.
- The KNN prefix artifact and the main training artifact were cache misses. Both completed Clang LLVM bitcode generation and `llc` PTX generation. The separate surrogate calibration artifact was a cache hit.
- The process exited 0 in 6.68 seconds. The epoch took 17.517 milliseconds, and MAE was finite at 1.7344.
- The reported native-artifact compile stall does not reproduce on current minimal.

## Change

This verification commit changes no tracked files because current behavior does not justify a production-code change.

</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.

**Comment by nm-z** [comment]
The public Recipe run on backend=nv0 remains active beyond 60 seconds before completing the one-epoch path. The earliest failure is native artifact compilation through `compile_model` -> `NativeModelIr::from_graph` -> NVIDIA LLVM/PTX compilation for the quantized `knn(3)` predictor and `perc(8)` block. This small workload should complete in under a few seconds; the dominant gap is the nv0 native compiler stall, not epoch arithmetic or data traffic.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `c269b4a6fb439487`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 325,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized knn/perc model with integer arithmetic",
  "body": "The public Recipe run on backend=nv0 remains active beyond 60 seconds before completing the one-epoch path. The earliest failure is native artifact compilation through `compile_model` -> `NativeModelIr::from_graph` -> NVIDIA LLVM/PTX compilation for the quantized `knn(3)` predictor and `perc(8)` block. This small workload should complete in under a few seconds; the dominant gap is the nv0 native compiler stall, not epoch arithmetic or data traffic.",
  "rationale": "Issue #325 and its comments represent the same earliest public failure and implementation path: a quantized knn predictor paired with a perceptron reaches the NVIDIA native artifact compiler before useful epoch execution. Source inspection confirms `.knn(3)` lowers through `fit_knn` and `lower_estimator` to `Primitive::Predictor`, while `.perc(8)` lowers through the perceptron projection path; `compile_model` emits the graph and invokes the NVIDIA LLVM/PTX assembler before the training loop. The records_json loader, bf16 arithmetic, activations, normalization, quantization variants, loss, split, and seed do not establish a different earliest cause. The measured 60 seconds is materially slower than the roughly sub-second to few-second workload justifies, so this is additional evidence for #325."
}
```

## Deterministic failure packet

kind=performance
id=c269b4a6fb439487
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:23712 next:23713 composition:118808094440928
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-8182-23712/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/categoric/records_json.json")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.knn(3).leak().norm(batch).iq(3).xs
		.perc(8).cos().iq(2).xxs
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(118808094440928)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.bf(16)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0
